File: tests/discord/router.test.ts
import { describe, expect, it } from "bun:test";
import type { Config } from "../../src/config/index.js";
import { getWebhookUrls } from "../../src/discord/router.js";

function createConfig(overrides: Partial<Config["discord"]> = {}): Config {
	const defaultWebhooks = {
		default: "https://discord.com/api/webhooks/default/token",
		info: "https://discord.com/api/webhooks/info/token",
		errors: "https://discord.com/api/webhooks/errors/token",
	};

	return {
		rabbitmq: { url: "amqp://localhost" },
		discord: {
			webhooks: overrides.webhooks ?? defaultWebhooks,
			routes: {
				ci: "info",
				pr: "info",
				...overrides.routes,
			},
			errorRoutes: overrides.errorRoutes ?? [
				"ci.failure",
				"deploy.failure",
				"dlq",
				"polling.failure",
			],
			defaultWebhook: overrides.defaultWebhook ?? "https://discord.com/api/webhooks/default/token",
		},
		loki: { host: undefined },
		logLevel: "info",
	};
}

describe("getWebhookUrls", () => {
	describe("primary webhook routing", () => {
		it("should route to mapped webhook when route exists", () => {
			// Arrange
			const config = createConfig();

			// Act
			const urls = getWebhookUrls("ci.success", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/info/token");
		});

		it("should route to default webhook when no route exists", () => {
			// Arrange
			const config = createConfig();

			// Act
			const urls = getWebhookUrls("unknown.event", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/default/token");
			expect(urls).toHaveLength(1);
		});

		it("should route to default when mapped webhook does not exist", () => {
			// Arrange
			const config = createConfig({
				routes: { ci: "nonexistent" },
			});

			// Act
			const urls = getWebhookUrls("ci.success", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/default/token");
		});
	});

	describe("error webhook routing", () => {
		it("should add errors webhook for error routes", () => {
			// Arrange
			const config = createConfig();

			// Act
			const urls = getWebhookUrls("ci.failure", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/info/token");
			expect(urls).toContain("https://discord.com/api/webhooks/errors/token");
			expect(urls).toHaveLength(2);
		});

		it("should add errors webhook for dlq routes", () => {
			// Arrange
			const config = createConfig();

			// Act
			const urls = getWebhookUrls("dlq.some-service", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/errors/token");
		});

		it("should not add errors webhook when not configured", () => {
			// Arrange
			const config = createConfig({
				webhooks: {
					default: "https://discord.com/api/webhooks/default/token",
					info: "https://discord.com/api/webhooks/info/token",
				},
			});

			// Act
			const urls = getWebhookUrls("ci.failure", config);

			// Assert
			expect(urls).toHaveLength(1);
		});

		it("should not add errors webhook for non-error routes", () => {
			// Arrange
			const config = createConfig();

			// Act
			const urls = getWebhookUrls("ci.success", config);

			// Assert
			expect(urls).not.toContain("https://discord.com/api/webhooks/errors/token");
			expect(urls).toHaveLength(1);
		});

		it("should not duplicate if primary and errors are same webhook", () => {
			// Arrange
			const config = createConfig({
				webhooks: {
					default: "https://discord.com/api/webhooks/default/token",
					info: "https://discord.com/api/webhooks/same/token",
					errors: "https://discord.com/api/webhooks/same/token",
				},
			});

			// Act
			const urls = getWebhookUrls("ci.failure", config);

			// Assert
			expect(urls).toHaveLength(1);
			expect(urls[0]).toBe("https://discord.com/api/webhooks/same/token");
		});
	});

	describe("routing key extraction", () => {
		it("should extract category from single-part routing key", () => {
			// Arrange
			const config = createConfig({ routes: { deploy: "info" } });

			// Act
			const urls = getWebhookUrls("deploy", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/info/token");
		});

		it("should extract category from multi-part routing key", () => {
			// Arrange
			const config = createConfig({ routes: { notifications: "info" } });

			// Act
			const urls = getWebhookUrls("notifications.dlq.some-service", config);

			// Assert
			expect(urls).toContain("https://discord.com/api/webhooks/info/token");
		});
	});
});

File: test/notification_service/discord/router_test.exs
defmodule NotificationService.Discord.RouterTest do
  use ExUnit.Case, async: true

  alias NotificationService.Discord.Router

  setup do
    base_webhooks = %{
      "default" => "https://discord.com/api/webhooks/default/token",
      "info" => "https://discord.com/api/webhooks/info/token",
      "prs" => "https://discord.com/api/webhooks/prs/token",
      "media" => "https://discord.com/api/webhooks/media/token",
      "alerts" => "https://discord.com/api/webhooks/alerts/token",
      "errors" => "https://discord.com/api/webhooks/errors/token"
    }

    base_routes = %{
      "ci" => "info",
      "pr" => "prs",
      "downloads" => "media",
      "deploy" => "info",
      "polling" => "alerts",
      "dlq" => "alerts"
    }

    base_error_routes = ["ci.failure", "deploy.failure", "dlq", "polling.failure"]

    Application.put_env(:notification_service, :discord_webhooks, base_webhooks)
    Application.put_env(:notification_service, :discord_routes, base_routes)
    Application.put_env(:notification_service, :discord_error_routes, base_error_routes)

    on_exit(fn ->
      Application.delete_env(:notification_service, :discord_webhooks)
      Application.delete_env(:notification_service, :discord_routes)
      Application.delete_env(:notification_service, :discord_error_routes)
    end)

    %{webhooks: base_webhooks, routes: base_routes, error_routes: base_error_routes}
  end

  describe "route/1" do
    test "routes CI messages to the info webhook" do
      urls = Router.route("ci.success")

      assert urls == ["https://discord.com/api/webhooks/info/token"]
    end

    test "routes PR messages to the prs webhook" do
      urls = Router.route("pr.opened")

      assert urls == ["https://discord.com/api/webhooks/prs/token"]
    end

    test "routes download messages to the media webhook" do
      urls = Router.route("downloads.complete")

      assert urls == ["https://discord.com/api/webhooks/media/token"]
    end

    test "routes deploy messages to the info webhook" do
      urls = Router.route("deploy.success")

      assert urls == ["https://discord.com/api/webhooks/info/token"]
    end

    test "routes polling messages to the alerts webhook" do
      urls = Router.route("polling.failure")

      assert "https://discord.com/api/webhooks/alerts/token" in urls
    end

    test "routes DLQ messages to the alerts webhook" do
      urls = Router.route("dlq.qbittorrent")

      assert "https://discord.com/api/webhooks/alerts/token" in urls
    end

    test "falls back to default webhook for unknown routing keys" do
      urls = Router.route("unknown.event")

      assert urls == ["https://discord.com/api/webhooks/default/token"]
    end

    test "falls back to default when routing key has no dots" do
      urls = Router.route("standalone")

      assert urls == ["https://discord.com/api/webhooks/default/token"]
    end

    test "adds errors webhook for ci.failure" do
      urls = Router.route("ci.failure")

      assert "https://discord.com/api/webhooks/info/token" in urls
      assert "https://discord.com/api/webhooks/errors/token" in urls
      assert length(urls) == 2
    end

    test "adds errors webhook for deploy.failure" do
      urls = Router.route("deploy.failure")

      assert "https://discord.com/api/webhooks/info/token" in urls
      assert "https://discord.com/api/webhooks/errors/token" in urls
    end

    test "adds errors webhook for DLQ alerts (prefix match)" do
      urls = Router.route("dlq.qbittorrent")

      assert "https://discord.com/api/webhooks/alerts/token" in urls
      assert "https://discord.com/api/webhooks/errors/token" in urls
    end

    test "adds errors webhook for polling.failure" do
      urls = Router.route("polling.failure")

      assert "https://discord.com/api/webhooks/alerts/token" in urls
      assert "https://discord.com/api/webhooks/errors/token" in urls
    end

    test "does not add errors webhook when errors webhook is not configured" do
      Application.put_env(
        :notification_service,
        :discord_webhooks,
        Map.delete(
          Application.get_env(:notification_service, :discord_webhooks),
          "errors"
        )
      )

      urls = Router.route("ci.failure")

      assert urls == ["https://discord.com/api/webhooks/info/token"]
    end

    test "deduplicates when errors webhook equals primary webhook" do
      Application.put_env(
        :notification_service,
        :discord_webhooks,
        Map.put(
          Application.get_env(:notification_service, :discord_webhooks),
          "errors",
          "https://discord.com/api/webhooks/alerts/token"
        )
      )

      urls = Router.route("dlq.qbittorrent")

      assert urls == ["https://discord.com/api/webhooks/alerts/token"]
    end

    test "falls back to default when webhook name does not exist" do
      Application.put_env(:notification_service, :discord_routes, %{"ci" => "nonexistent"})

      urls = Router.route("ci.success")

      assert urls == ["https://discord.com/api/webhooks/default/token"]
    end

    test "does not duplicate ci.success error route (no error route match)" do
      urls = Router.route("ci.success")

      assert urls == ["https://discord.com/api/webhooks/info/token"]
    end
  end
end
